File: simnibs/examples/simulations/TI.m
%
% example script that runs two simnibs tDCS simulations
% and calculates maximal amplitude of the TI envelope from the E-fields 
% 
% A. Thielscher, 2020
%

%% General information

S = sim_struct('SESSION');
S.fnamehead = 'ernie.msh'; % head mesh
S.pathfem = 'TI'; %Folder for the simulation output

%% Define tDCS simulation: First electrode pair
S.poslist{1} = sim_struct('TDCSLIST');
S.poslist{1}.currents = [0.001, -0.001]; % Current flow though each channel (mA)

%First Electrode
S.poslist{1}.electrode(1).channelnr = 1; % Connect the electrode to the first channel
S.poslist{1}.electrode(1).centre = 'F5'; % Place it over F5
S.poslist{1}.electrode(1).shape = 'ellipse'; %round electrodes
S.poslist{1}.electrode(1).dimensions = [40, 40]; % Dimension in mm
S.poslist{1}.electrode(1).thickness = 2; % 4 mm thickness

%Second Electrode
S.poslist{1}.electrode(2).channelnr = 2;
S.poslist{1}.electrode(2).centre = 'P5';
S.poslist{1}.electrode(2).shape = 'ellipse';
S.poslist{1}.electrode(2).dimensions = [40, 40];
S.poslist{1}.electrode(2).thickness = 2;

%% Define second electrode pair: identical, except for electrode positions
S.poslist{2} = S.poslist{1};

%First Electrode
S.poslist{2}.electrode(1).centre = 'F6'; % update Place it over F6
S.poslist{2}.electrode(2).centre = 'P6';


%% Run Simulation
run_simnibs(S);

%% analyse simulations
m1 = mesh_load_gmsh4(fullfile(S.pathfem, 'ernie_TDCS_1_scalar.msh'));
m2 = mesh_load_gmsh4(fullfile(S.pathfem, 'ernie_TDCS_2_scalar.msh'));

% remove all tetrahedra and triangles belonging to the electrodes so that
% the two meshes have same number of elements
m1=mesh_extract_regions(m1, 'region_idx', [1:99 1001:1099]);
m2=mesh_extract_regions(m2, 'region_idx', [1:99 1001:1099]);

% calculate the maximal amplitude of the TI envelope for all elements
maxTI=get_maxTI( m1.element_data{get_field_idx(m1,'E','elem')}, ...
                 m2.element_data{get_field_idx(m2,'E','elem')} );

% make a new mesh that contains the field strengths and the 
% amplitude of the TI envelope for visualization
mout = m1;
mout.element_data={};
mout.element_data{1} = m1.element_data{get_field_idx(m1,'normE','elem')};
mout.element_data{1}.name = 'normE_1';
mout.element_data{2} = m2.element_data{get_field_idx(m2,'normE','elem')};
mout.element_data{2}.name = 'normE_2';
mout.element_data{3} = maxTI;

% save for visualization in gmsh
mesh_save_gmsh4(mout, fullfile(S.pathfem, 'ernie_TIenvelope.msh'));

% show field strengths and amplitude of TI envelope on GM surface
mesh_show_surface(mout,'field_idx','normE_1');
mesh_show_surface(mout,'field_idx','normE_2');
mesh_show_surface(mout,'field_idx','maxTIamplitude');

File: simnibs/examples/simulations/TI.py
# -*- coding: utf-8 -*-
"""
 example script that runs two simnibs tDCS simulations
 and calculates maximal amplitude of the TI envelope from the E-fields
 
 Created on Thu Jun 23 17:41:21 2022

@author: axthi
"""

from copy import deepcopy
import os
import numpy as np

from simnibs import sim_struct, run_simnibs, mesh_io, ElementTags
from simnibs.utils import TI_utils as TI


"""
     set up and run simulations for the two electrode pairs
"""

# specify general parameters
S = sim_struct.SESSION()
S.subpath = 'm2m_ernie'  # m2m-folder of the subject
S.pathfem = 'TI'  # Directory for the simulation

# specify first electrode pair
tdcs = S.add_tdcslist()
tdcs.currents = [0.001, -0.001]  # Current flow though each channel (A)

electrode = tdcs.add_electrode()
electrode.channelnr = 1
electrode.centre = 'F5'  
electrode.shape = 'ellipse' 
electrode.dimensions = [40, 40]  # diameter in [mm]
electrode.thickness = 2  # 2 mm thickness

electrode = tdcs.add_electrode()
electrode.channelnr = 2
electrode.centre = 'P5'
electrode.shape = 'ellipse'
electrode.dimensions = [40, 40]
electrode.thickness = 2

# specify second electrode pair
tdcs = S.add_tdcslist(deepcopy(tdcs))
tdcs.electrode[0].centre = 'F6' 
tdcs.electrode[1].centre = 'P6' 

run_simnibs(S)


"""
    generate the TI field from the simulation results
"""
m1 = mesh_io.read_msh(os.path.join(S.pathfem, 'ernie_TDCS_1_scalar.msh'))
m2 = mesh_io.read_msh(os.path.join(S.pathfem, 'ernie_TDCS_2_scalar.msh'))

# remove all tetrahedra and triangles belonging to the electrodes so that
# the two meshes have same number of elements
tags_keep = np.hstack((np.arange(ElementTags.TH_START, ElementTags.SALINE_START - 1), np.arange(ElementTags.TH_SURFACE_START, ElementTags.SALINE_TH_SURFACE_START - 1)))
m1=m1.crop_mesh(tags = tags_keep)
m2=m2.crop_mesh(tags = tags_keep)

# calculate the maximal amplitude of the TI envelope
ef1=m1.field['E']
ef2=m2.field['E']
TImax = TI.get_maxTI(ef1.value, ef2.value)

# make a new mesh for visualization of the field strengths
# and the amplitude of the TI envelope
mout = deepcopy(m1)
mout.elmdata = []
mout.add_element_field(ef1.norm(), 'magnE - pair 1')
mout.add_element_field(ef2.norm(), 'magnE - pair 2')                    
mout.add_element_field(TImax,'TImax')
mesh_io.write_msh(mout,os.path.join(S.pathfem, 'TI.msh'))
v = mout.view(
    visible_tags=[1002, 1006],
    visible_fields='TImax',    
    )
v.write_opt(os.path.join(S.pathfem, 'TI.msh'))
mesh_io.open_in_gmsh(os.path.join(S.pathfem, 'TI.msh'), True)
            
            

